### PR TITLE
Ember Paper: Implement text input

### PR DESCRIPTION
**Pull request by adiehl96**
Follow up to pr #66

### CODE REVIEW

**Comment by adiehl96** [comment]
I addressed the requested changes from #66

File: addon/components/paper-text-field.js [hunk 1]
```diff
@@ -0,0 +1,76 @@
+import Component        from 'ember-component';
+import formFieldOptions from
+  'ember-railio-form-components/mixins/form-field-options';
+import get   from 'ember-metal/get';
+import set   from 'ember-metal/set';
+
+export default Component.extend(formFieldOptions, {
+
```
Comment (michaelw8) [inline addon/components/paper-text-field.js]: Unnecessary empty line

File: addon/components/paper-text-field.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+      this.send('changed', _value);
+    },
+
+    changed(value) {
+      if (typeof this.attrs.updated === 'function') {
+        this.attrs.updated(value);
+      }
+    }
+  }
+});
```
Comment (michaelw8) [inline addon/components/paper-text-field.js]: I guess you just copied the whole file? Did you reconsider if this is the best way?

File: addon/components/paper-text-field.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+    if (typeof get(this, 'format') === 'function') {
+      _value = this.format(_value);
+      set(this, '_value', _value);
+    }
+
+    this.send('changed', _value);
+  },
+
+  keyUp(e) {
```
Comment (michaelw8) [inline addon/components/paper-text-field.js]: Why not just bind these functions to the component (just like we discussed earlier).

File: addon/components/paper-text-field.js [hunk 4]
```diff
@@ -0,0 +1,76 @@
+import Component        from 'ember-component';
+import formFieldOptions from
+  'ember-railio-form-components/mixins/form-field-options';
+import get   from 'ember-metal/get';
+import set   from 'ember-metal/set';
+
+export default Component.extend(formFieldOptions, {
+
+  focusIn() {
```
Comment (michaelw8) [inline addon/components/paper-text-field.js]: (see next comment)

File: addon/components/paper-text-field.js [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+    if (e.keyCode === 27) {
+      if (typeof this.attrs.escape === 'function') {
+        this.attrs.escape();
+      }
+    }
+  },
+
+  didReceiveAttrs() {
+    let value = this.getAttr('value');
```
Comment (michaelw8) [inline addon/components/paper-text-field.js]: getAttr?

File: addon/components/paper-text-field.js [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        if (originalValue !== _value) {
+          set(this, '_value', _value);
+        }
+      }
+
+      this.send('changed', _value);
+    },
+
+    changed(value) {
+      if (typeof this.attrs.updated === 'function') {
```
Comment (michaelw8) [inline addon/components/paper-text-field.js]: I think you get an error when updated is undefined, what would make this if-statement unnecessary

File: app/templates/components/paper-text-field.hbs [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+              placeholder=placeholder
+              type=type
+              disabled=disabled
+              autofocus=autofocus
+              onfocus=onFocus
+              onblur=onBlur
+              onkeydown=onKeyDown
+              onkeyup=onKeyUp
+              onclick=onClick
+              oninput=oninput
```
[the remark this answers is not in the document]
Comment (michaelw8) [reply app/templates/components/paper-text-field.hbs]: Why do you need this? Should this belong to the passThru?

File: tests/integration/components/paper-text-field-test.js [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    await render(hbs`{{paper-text-field value=value}}`);
+
+    assert.equal(find('input').value, 'testing');
+  });
+
+  test('changing value changes input text', async function(assert) {
+    this.set('value', 'testing');
+    await render(hbs`{{paper-text-field value=value}}`);
+
+    run(() => {
```
Comment (michaelw8) [inline tests/integration/components/paper-text-field-test.js]: I think you don't need this. Maybe replace it with `await`

File: tests/integration/components/paper-text-field-test.js [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+
+    this.set('value', '');
+    this.actions.update = function(value) {
+      assert.equal(value, 'x', 'calls update function with new value');
+    };
+
+    await render(hbs`{{paper-text-field value=value updated=(action "update")}}`);
+
+    let $input = this.$('input');
+    run(() => {
```
Comment (michaelw8) [inline tests/integration/components/paper-text-field-test.js]: same

File: addon/components/paper-text-field.js (hunk 2 above)
Comment (adiehl96) [reply addon/components/paper-text-field.js]: I tried to be consistent with the way things were done before, so everything that is there now is needed for other components that extend the text field.

File: addon/components/paper-text-field.js (hunk 3 above)
Comment (adiehl96) [reply addon/components/paper-text-field.js]: I think we discussed this, while you were sitting next to me

File: addon/components/paper-text-field.js (hunk 4 above)
Comment (adiehl96) [reply addon/components/paper-text-field.js]: (see next comment)

File: addon/components/paper-text-field.js (hunk 6 above)
Comment (adiehl96) [reply addon/components/paper-text-field.js]: You only get that error when using form-field, but we sometimes just use text-fields standalone, which means they need to check themselves.

File: addon/components/paper-text-field.js (hunk 5 above)
Comment (adiehl96) [reply addon/components/paper-text-field.js]: this is an ember invention to also access attributes, I didn't change it

**Comment by adiehl96** [comment]
Ready for review

File: tests/integration/components/paper-text-field-test.js [hunk 10]
```diff
@@ -0,0 +1,93 @@
+import { module, test } from 'qunit';
+import { setupRenderingTest } from 'ember-qunit';
+import { render, find, blur, triggerEvent, focus } from '@ember/test-helpers';
+import hbs from 'htmlbars-inline-precompile';
```
Comment (michaelw8) [inline tests/integration/components/paper-text-field-test.js]: Alignment

File: tests/integration/components/paper-text-field-test.js [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+
+  hooks.beforeEach(function() {
+    this.actions = {};
+    this.send = (actionName, ...args) => this.actions[actionName].apply(this, args);
+  });
+
+  test('renders input with placeholder', async function(assert) {
+    await render(hbs`{{paper-text-field placeholder='Type your value here'}}`);
+
+    assert.equal(this.$('input')[0].getAttribute('placeholder'),
```
Comment (michaelw8) [inline tests/integration/components/paper-text-field-test.js]: Can you do this without jquery?

File: tests/integration/components/paper-text-field-test.js [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+    assert.expect(2);
+
+    this.set('value', '');
+    this.actions.update = function(value) {
+      assert.equal(value, 'x', 'calls update function with new value');
+    };
+
+    await render(hbs`{{paper-text-field value=value updated=(action "update")}}`);
+
+    let $input = this.$('input');
```
Comment (michaelw8) [inline tests/integration/components/paper-text-field-test.js]: No jquery

File: tests/integration/components/paper-text-field-test.js [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+import { setupRenderingTest } from 'ember-qunit';
+import { render, find, blur, triggerEvent, focus } from '@ember/test-helpers';
+import hbs from 'htmlbars-inline-precompile';
+
+module('Integration | Component | {{paper-text-field}}', function(hooks) {
+  setupRenderingTest(hooks);
+
+  hooks.beforeEach(function() {
+    this.actions = {};
+    this.send = (actionName, ...args) => this.actions[actionName].apply(this, args);
```
Comment (michaelw8) [inline tests/integration/components/paper-text-field-test.js]: `this.actions = { updated: () => {} };`? Like you in select-auto-complete-test

**Pull request merged**